### PR TITLE
[KYUUBI #6618] Support http bearer token authentication for REST protocol

### LINKED ISSUE #6590: [FEATURE] Support bearer token authentication
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/kyuubi/issues?q=is%3Aissue) and found no similar issues.

### Describe the feature

The same with celeborn https://github.com/apache/celeborn/blob/main/service/src/main/scala/org/apache/celeborn/server/common/http/authentication/BearerAuthenticationHandler.scala

To support bearer token authentication.

### Motivation

Bearer token authentication is widely used.

### Describe the solution

Refer: https://github.com/apache/celeborn/blob/main/service/src/main/scala/org/apache/celeborn/server/common/http/authentication/BearerAuthenticationHandler.scala
https://github.com/apache/celeborn/pull/2440

### Additional context

_No response_

### Are you willing to submit PR?

- [ ] Yes. I would be willing to submit a PR with guidance from the Kyuubi community to improve.
- [ ] No. I cannot submit a PR at this time.

### LINKED ISSUE #6618: [KYUUBI #6590][Subtask] Support http bearer authentication for REST protocol
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/kyuubi/issues) and found no similar issues.

### Describe the subtask

Support http bearer authentication for REST protocol

### Parent issue

KYUUBI #6590

### Are you willing to submit PR?

- [ ] Yes. I would be willing to submit a PR with guidance from the Kyuubi community to improve.
- [ ] No. I cannot submit a PR at this time.

### PR DESCRIPTION
**Pull request by George314159**
# :mag: Description
## Issue References 🔗

This pull request fixes #6618  

## Describe Your Solution 🔧

It is a subtask of #6590 
This PR is to support http bearer token authentication for REST protocol. In addition to BasicAuthenticationHandler, BearerAuthenticationHandler will be added to handle http bear token authentication. They will both support CUSTOM AuthType. In order to distinguish them, two new configurations are added: kyuubi.authentication.custom.basic.class and kyuubi.authentication.custom.bearer.class. For http bear token custom authentication, users could implement the new 'org.apache.kyuubi.service.authentication.TokenAuthenticationProvider', and specify it in the configuration.

## Types of changes :bookmark:

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

#### Related Unit Tests

---

# Checklist 📝

- [x] This patch was not authored or co-authored using [Generative Tooling](https://www.apache.org/legal/generative-tooling.html)

**Be nice. Be informative.**

### CODE REVIEW

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 1]
```diff
@@ -823,6 +824,11 @@ object KyuubiConf {
       "          <li><code>KERBEROS</code></li>" +
       "        </ul>" +
       "      </li>" +
+      "      <li>SASL/OAUTHBEARER" +
+      "        <ul>" +
+      "          <li><code>BEARER</code></li>" +
+      "        </ul>" +
+      "      </li>" +
       "    </ul>" +
```
Comment (turboFei) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:832]: Hi @George314159 thanks for raising this PR.

Some comments, for current design:
we do not need to add new item into kyuubi.authentication, for example, now I set `kyuubi.authentication=KERBEROS,CUSTOM` in server end, for `Bearer` authentication, it belongs to `CUSTOM` authentication.

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 2]
```diff
@@ -839,7 +845,8 @@ object KyuubiConf {
   val AUTHENTICATION_CUSTOM_CLASS: OptionalConfigEntry[String] =
     buildConf("kyuubi.authentication.custom.class")
       .doc("User-defined authentication implementation of " +
-        "org.apache.kyuubi.service.authentication.PasswdAuthenticationProvider")
+        "org.apache.kyuubi.service.authentication.PasswdAuthenticationProvider " +
+        "or org.apache.kyuubi.service.authentication.TokenAuthenticationProvider")
```
Comment (turboFei) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:849]: we shall involve two new config items.

`kyuubi.authentication.custom.basic.class` and `kyuubi.authentication.custom.bearer.class`

Comment (turboFei) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:849]: And for 
`kyuubi.authentication.custom.basic.class`, it should fallback to `AUTHENTICATION_CUSTOM_CLASS`

.fallbackConf(AUTHENTICATION_CUSTOM_CLASS)

File: kyuubi-common/src/main/scala/org/apache/kyuubi/service/authentication/AuthTypes.scala [hunk 3]
```diff
@@ -20,5 +20,5 @@ package org.apache.kyuubi.service.authentication
 object AuthTypes extends Enumeration {
   type AuthType = Value
 
-  val NOSASL, NONE, LDAP, JDBC, KERBEROS, CUSTOM = Value
+  val NOSASL, NONE, LDAP, JDBC, KERBEROS, CUSTOM, BEARER = Value
```
Comment (turboFei) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/service/authentication/AuthTypes.scala:23]: not needed

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/http/authentication/AuthSchemes.scala [hunk 4]
```diff
@@ -20,5 +20,5 @@ package org.apache.kyuubi.server.http.authentication
 object AuthSchemes extends Enumeration {
   type AuthScheme = Value
 
-  val BASIC, NEGOTIATE, KYUUBI_INTERNAL = Value
+  val BASIC, NEGOTIATE, BEARER, KYUUBI_INTERNAL = Value
```
Comment (turboFei) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/http/authentication/AuthSchemes.scala:23]: val BASIC, BEARER, NEGOTIATE, KYUUBI_INTERNAL = Value

**Comment by George314159** [comment]
Thanks for review, updated the design based on comments.

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 5]
```diff
@@ -845,6 +845,23 @@ object KyuubiConf {
       .stringConf
       .createOptional
 
+  val AUTHENTICATION_CUSTOM_BASIC_CLASS: ConfigEntry[Option[String]] =
+    buildConf("kyuubi.authentication.custom.basic.class")
+      .doc("User-defined Basic authentication implementation of " +
+        "org.apache.kyuubi.service.authentication.PasswdAuthenticationProvider")
```
Comment (turboFei) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:851]: add `for http basic authentication` to make it clear.

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 6]
```diff
@@ -845,6 +845,23 @@ object KyuubiConf {
...
+      .version("1.6.0")
+      .serverOnly
+      .fallbackConf(AUTHENTICATION_CUSTOM_CLASS)
+
+  val AUTHENTICATION_CUSTOM_BEARER_CLASS: OptionalConfigEntry[String] =
+    buildConf("kyuubi.authentication.custom.bearer.class")
+      .doc("User-defined Token authentication implementation of " +
+        "org.apache.kyuubi.service.authentication.TokenAuthenticationProvider")
+      .version("1.6.0")
```
Comment (turboFei) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:860]: add `for http bearer authentication` to make it clear.

Comment (turboFei) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:860]: 1.10.0

File: kyuubi-common/src/main/scala/org/apache/kyuubi/service/authentication/AuthenticationProviderFactory.scala [hunk 7]
```diff
@@ -55,6 +55,32 @@ object AuthenticationProviderFactory {
...
+          "authentication.custom.basic.class must be set when auth method was CUSTOM.")
+      }
+      ClassUtils.createInstance(className.get, classOf[PasswdAuthenticationProvider], conf)
+    case _ => throw new AuthenticationException("Not a valid authentication method")
+  }
+
+  def getTokenAuthenticationProvider(
+      providerClass: String,
+      conf: KyuubiConf): TokenAuthenticationProvider = {
+    if (providerClass.isEmpty) {
```
Comment (turboFei) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/service/authentication/AuthenticationProviderFactory.scala:77]: if (StringUtils.isBlank(providerClass))

File: kyuubi-common/src/main/scala/org/apache/kyuubi/service/authentication/AuthenticationProviderFactory.scala [hunk 8]
```diff
@@ -55,6 +55,32 @@ object AuthenticationProviderFactory {
...
+      throw new AuthenticationException(
+        "authentication.custom.bearer.class must be set when auth method was BEARER.")
```
Comment (turboFei) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/service/authentication/AuthenticationProviderFactory.scala:79]: `authentication.custom.bearer.class must be set for token authentication`.

File: kyuubi-common/src/main/scala/org/apache/kyuubi/service/authentication/EngineSecureAuthenticationProviderImpl.scala [hunk 9]
```diff
@@ -17,8 +17,12 @@
...
 
-class EngineSecureAuthenticationProviderImpl extends PasswdAuthenticationProvider {
+class EngineSecureAuthenticationProviderImpl extends PasswdAuthenticationProvider
+  with TokenAuthenticationProvider {
   override def authenticate(user: String, password: String): Unit = {
     InternalSecurityAccessor.get().authToken(password)
   }
+  override def authenticate(token: String): Unit = {
+    InternalSecurityAccessor.get().authToken(token)
+  }
```
Comment (turboFei) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/service/authentication/EngineSecureAuthenticationProviderImpl.scala:27]: not needed for `EngineSecureAuthenticationProviderImpl`.

EngineSecureAuthenticationProviderImpl only for `PasswdAuthenticationProvider`.

File: kyuubi-common/src/main/scala/org/apache/kyuubi/service/authentication/TokenAuthenticationProvider.scala [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+   * to authenticate users for their requests.
+   * If the token is to be granted, return nothing/throw nothing.
+   * When the token is to be disallowed, throw an appropriate [[AuthenticationException]].
+   *
+   * @param token The token received over the connection request
+   *
+   * @throws AuthenticationException When the token is found to be invalid by the implementation
+   */
+  @throws[AuthenticationException]
+  def authenticate(token: String): Unit
```
Comment (turboFei) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/service/authentication/TokenAuthenticationProvider.scala:35]: def authenticate(credential: TokenCredential): Principal

Comment (turboFei) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/service/authentication/TokenAuthenticationProvider.scala:35]: Please follow the same pattern likes https://github.com/apache/celeborn/blob/main/spi/src/main/java/org/apache/celeborn/spi/authentication/TokenAuthenticationProvider.java.

**Comment by turboFei** [comment]
Please also add UT in `KyuubiRestAuthenticationSuite`.

**Review by turboFei — APPROVED** [review]
LGTM overall

**Comment by turboFei** [comment]
could you update the PR description for more details? @George314159

**Comment by turboFei** [comment]
also cc @pan3793

**Comment by turboFei** [comment]
For ThriftHttp protocol, https://github.com/apache/kyuubi/pull/6621

File: kyuubi-common/src/main/scala/org/apache/kyuubi/service/authentication/AuthenticationProviderFactory.scala [hunk 11]
```diff
@@ -62,4 +64,30 @@ object AuthenticationProviderFactory {
...
+      conf: KyuubiConf): PasswdAuthenticationProvider = method match {
+    case AuthMethods.NONE => new AnonymousAuthenticationProviderImpl
+    case AuthMethods.LDAP => new LdapAuthenticationProviderImpl(conf)
+    case AuthMethods.JDBC => new JdbcAuthenticationProviderImpl(conf)
+    case AuthMethods.CUSTOM =>
+      val className = conf.get(KyuubiConf.AUTHENTICATION_CUSTOM_BASIC_CLASS)
+      if (className.isEmpty) {
+        throw new AuthenticationException(
+          "authentication.custom.basic.class must be set for http basic authentication.")
+      }
```
Comment (pan3793) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/service/authentication/AuthenticationProviderFactory.scala:79]: this is caused by admin's wrong configuration, should not throw `AuthenticationException`
```
require(className.nonEmpty, "xxxxx")
```

please correct the mentioned configuration key
`authentication.custom.basic.class` => `kyuubi.authentication.custom.basic.class`

File: kyuubi-common/src/main/scala/org/apache/kyuubi/service/authentication/AuthenticationProviderFactory.scala [hunk 12]
```diff
@@ -62,4 +64,30 @@ object AuthenticationProviderFactory {
...
+      ClassUtils.createInstance(className.get, classOf[PasswdAuthenticationProvider], conf)
+    case _ => throw new AuthenticationException("Not a valid authentication method")
+  }
+
+  def getHttpBearerAuthenticationProvider(
+      providerClass: String,
+      conf: KyuubiConf): TokenAuthenticationProvider = {
+    if (StringUtils.isBlank(providerClass)) {
+      throw new AuthenticationException(
+        "authentication.custom.bearer.class must be set for http bearer authentication.")
```
Comment (pan3793) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/service/authentication/AuthenticationProviderFactory.scala:89]: ditto

File: kyuubi-common/src/test/scala/org/apache/kyuubi/service/authentication/CustomAuthenticationProviderImplSuite.scala [hunk 13]
```diff
@@ -21,7 +21,7 @@ import javax.security.sasl.AuthenticationException
 
 import org.apache.kyuubi.KyuubiFunSuite
 import org.apache.kyuubi.config.KyuubiConf
-import org.apache.kyuubi.service.authentication.AuthenticationProviderFactory.getAuthenticationProvider
+import org.apache.kyuubi.service.authentication.AuthenticationProviderFactory.{getAuthenticationProvider, getHttpBearerAuthenticationProvider}
 
 class CustomAuthenticationProviderImplSuite extends KyuubiFunSuite {
   test("Test user defined authentication") {
```
Comment (turboFei) [inline kyuubi-common/src/test/scala/org/apache/kyuubi/service/authentication/CustomAuthenticationProviderImplSuite.scala:27]: this UT also need to be fixed

**Comment by pan3793** [comment]
LGTM

**Review by turboFei — APPROVED** [review]
LGTM

**Comment by pan3793** [comment]
Thanks, merged to master

**Pull request closed without merging**